tests/parts/test_document.py

```diff
@@ -45,6 +45,17 @@ def it_can_get_or_add_an_image(self, get_image_fixture):
         document_part.relate_to.assert_called_once_with(image_part_, RT.IMAGE)
         assert (rId, image) == (rId_, image_)
 
+    def it_can_iterate_the_story_parts(self, iter_parts_related_by_):
+        document_part = DocumentPart(None, None, None, None)
+        story_part_types = {RT.HEADER, RT.FOOTER, RT.FOOTNOTES, RT.ENDNOTES,
+                            RT.COMMENTS}
+
+        story_parts = document_part.iter_story_parts()
+
+        iter_parts_related_by_.assert_called_once_with(story_part_types)
+        for story_part in story_parts:
+            assert isinstance(story_part, DocumentPart)
```

Comment:
So this test is not sufficient to properly exercise the `.iter_story_parts()` method. Because the `.iter_parts_related_by()` method mock corresponds to an empty iterator, all you're testing is that the `(self,),` portion inside the `chain()` expression is present. If you move lines 96-98 above the return call (where they contribute nothing to the return value), the test still passes.

So you'll need to assign a return value to the `iter_parts_related_by_()` mock, perhaps something like:

```python
iter_parts_related_by_.return_value = iter((header_part_, footer_part_))
...
assert list(story_part_iter) == [document_part, header_part_, footer_part_]
```

Reply:
Fixed, added two mocks.